src/rpc/rpc_payment.cpp

```diff
@@ -0,0 +1,354 @@
+      hashing_blob = get_block_hashing_blob(info.block);
+      info.previous_top = info.top;
+      info.previous_payments = info.payments;
+      info.payments.clear();
+      info.hashing_blob = hashing_blob;
+      info.block_template_update_time = now;
+    }
+    info.top = top;
+    info.update_time = now;
+    hashing_blob = info.hashing_blob;
+    diff = m_diff;
+    credits_per_hash_found = m_credits_per_hash_found;
+    credits = info.credits;
+    return true;
+  }
+
+  bool rpc_payment::submit_nonce(const crypto::public_key &client, uint32_t nonce, const crypto::hash &top, int64_t &error_code, std::string &error_message, uint64_t &credits, crypto::hash &hash, cryptonote::block &block)
+  {
+    client_info &info = m_client_info[client]; // creates if not found
+    MINFO("client " << client << " sends nonce: " << nonce);
+    const bool is_current = top == info.top;
+    std::unordered_set<uint64_t> &payments = is_current ? info.payments : info.previous_payments;
+    if (payments.find(nonce) != payments.end())
+    {
+      MWARNING("Duplicate nonce " << nonce << " from " << (is_current ? "current" : "previous"));
+      ++m_nonces_dupe;
+      ++info.nonces_dupe;
+      info.credits = std::max(info.credits, PENALTY_FOR_DUPLICATE * m_credits_per_hash_found) - PENALTY_FOR_DUPLICATE * m_credits_per_hash_found;
+      error_code = CORE_RPC_ERROR_CODE_DUPLICATE_PAYMENT;
+      error_message = "Duplicate payment";
+      return false;
+    }
+    payments.insert(nonce);
+
+    if (info.hashing_blob.size() < 43)
+    {
+      // not initialized ?
+      error_code = CORE_RPC_ERROR_CODE_WRONG_BLOCKBLOB;
+      error_message = "not initialized";
+      return false;
+    }
+
+    const uint64_t now = time(NULL);
+    if (!is_current)
+    {
+      if (now > info.update_time + STALE_THRESHOLD)
+      {
+        MWARNING("Nonce is stale (top " << top << ", should be " << info.top << " or within " << STALE_THRESHOLD << " seconds");
+        ++m_nonces_stale;
+        ++info.nonces_stale;
+        info.credits = std::max(info.credits, PENALTY_FOR_STALE * m_credits_per_hash_found) - PENALTY_FOR_STALE * m_credits_per_hash_found;
+        error_code = CORE_RPC_ERROR_CODE_STALE_PAYMENT;
+        error_message = "stale payment";
+        return false;
+      }
+    }
+
+    cryptonote::blobdata hashing_blob = info.hashing_blob;
+    *(uint32_t*)(hashing_blob.data() + 39) = SWAP32LE(nonce);
+    const int cn_variant = hashing_blob[0] >= 7 ? hashing_blob[0] - 6 : 0;
+    crypto::cn_slow_hash(hashing_blob.data(), hashing_blob.size(), hash, cn_variant);
+    if (!check_hash(hash, m_diff))
+    {
+      MWARNING("Payment too low");
+      ++m_nonces_bad;
+      ++info.nonces_bad;
+      error_code = CORE_RPC_ERROR_CODE_PAYMENT_TOO_LOW;
+      error_message = "Payment too low";
+      info.credits = std::max(info.credits, PENALTY_FOR_BAD_HASH * m_credits_per_hash_found) - PENALTY_FOR_BAD_HASH * m_credits_per_hash_found;
+      return false;
+    }
```

Comment:
Maybe I miss something here, but right now I don't understand why a nonce that is not correct is reported as a payment that is too low. If that's done merely to avoid yet another extra error code, I am not sure it's a good idea: Maybe people should get reported quite clearly that they submit bad nonces.

Reply:
That's the same thing, isn't it ? If you send a hash below the threshold, it's bad.

Reply:
With "conventional" standalone miners, there is a danger that your miner does not use the correct PoW algorithm and calculates completely wrong hashes because of this, so even if a hash has "the right number of zeroes in the right place" it's still not valid. I was worrying about that happening and people not finding the problem because the message "Payment too low" does not give a hint into the right direction.

But maybe this whole scenario is moot because we use Monero's own PoW code that will hardly ever produce completely wrong hashes.

Reply:
OK, I'll change the message to give hints.